<a href="https://colab.research.google.com/github/28-Harshali/Housing-Prediction/blob/main/ML_LAB_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Harshali Bhoye

21102A0040

BE CMPN A

ML LAB 2

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('train.csv')

# Display the first few rows of the dataframe
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
df.info()
# Data structure and information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [ ]:
# Summary statistics
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [ ]:
df.isnull().sum()
# Check for missing values

Survived      0
Pclass        0
Sex           0
Age           0
SibSp         0
Parch         0
Fare          0
Embarked_Q    0
Embarked_S    0
dtype: int64

In [ ]:
df['Age'].fillna(df['Age'].median(), inplace=True)
# Fill missing values in 'Age' with the median

In [ ]:
df['Embarked'].fillna(df['Embarked'].mode()[0], inplace=True)
# Fill missing values in 'Embarked' with the mode

In [ ]:
df.drop(columns=['Cabin'], inplace=True)
# Drop the 'Cabin' column due to too many missing values

In [ ]:
df['Sex'] = df['Sex'].map({'male': 0, 'female': 1})
# Convert 'Sex' to numerical

In [ ]:
df = pd.get_dummies(df, columns=['Embarked'], drop_first=True) # One-hot encoding for 'Embarked'

In [ ]:
df.drop(columns=['PassengerId', 'Name', 'Ticket'], inplace=True) # Drop columns that are not needed for prediction

In [ ]:
X = df.drop(columns=['Survived'])
y = df['Survived'] # Define features (X) and target variable (y)

In [ ]:
from sklearn.model_selection import train_test_split
# Split the data into training and testing sets (80-20 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
from sklearn.linear_model import LogisticRegression

# Initialize the logistic regression model
model = LogisticRegression(max_iter=200)

# Train the model on the training data
model.fit(X_train, y_train)

LogisticRegression(max_iter=200)

In [ ]:
y_pred = model.predict(X_test) # Make predictions on the test set

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Generate confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print(f'Confusion Matrix:\n{conf_matrix}')

# Generate classification report
class_report = classification_report(y_test, y_pred)
print(f'Classification Report:\n{class_report}')


Accuracy: 0.81
Confusion Matrix:
[[90 15]
 [19 55]]
Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.86      0.84       105
           1       0.79      0.74      0.76        74

    accuracy                           0.81       179
   macro avg       0.81      0.80      0.80       179
weighted avg       0.81      0.81      0.81       179



In [ ]:
# Get feature importance (coefficients)
coefficients = pd.DataFrame({'Feature': X.columns, 'Coefficient': model.coef_[0]})
coefficients = coefficients.sort_values(by='Coefficient', ascending=False)
print(coefficients)


      Feature  Coefficient
1         Sex     2.590267
5        Fare     0.002577
2         Age    -0.030555
4       Parch    -0.107684
6  Embarked_Q    -0.112306
3       SibSp    -0.295076
7  Embarked_S    -0.399477
0      Pclass    -0.938107


# Analysis of Results
The model's accuracy is 0.81%. The confusion matrix shows the breakdown of true positives, false positives, true negatives, and false negatives. The classification report provides precision, recall, and F1-score for both classes.

# Feature Importance
The coefficients indicate that Sex, Pclass, and Embarked_S are the most influential features in predicting survival.

# Potential Improvements
To improve the model, we can:
- Use feature engineering to create new features.
- Try different models and compare their performance.
- Use cross-validation to better evaluate the model.